### PR TITLE
R2dbc support

### PR DESCRIPTION
**Pull request by Humblehound**
I've recently decided to try R2DBC and noticed that TestContainers is currently only compatible with JDBC. I've decided to add the support for it, however there are lots of impactful decisions to be made so I'd like some feedback on the assumptions I've made. I've currently implemented a very basic support for Postgresql containers. Before I go any further I'd like to discuss the following topics:

- Packaging. So far it seems possible to reuse `GenericContainer`, so I've decided to work with that and created a new package equivalent to the `org.testcontainers.jdbc`. The existing container implementations are tightly coupled with JDBC, so it would be best to have a separate class for each R2DBC container. The container implementations would be prefixed with `R2dbc`  and reside in the packages next to original JDBC versions.  

- JDBC URI. As @mp911de has mentioned [here](https://github.com/testcontainers/testcontainers-java/issues/1003) and also [here](https://github.com/r2dbc/r2dbc-spi/issues/10) it would be desirable to not reimplement DataSource discovery based on string parsing. For the time being, I suggest a following approach:
   - A database container can only be created explicitly in a test
   - A container should expose an R2dbc object so that a test client can easily get access to the DB without having to create it on their own. We could also expose just the `ConnectionFactory`, but so far I see no benefit in this. 
  - A container should also expose all the connection params (as a POJO or getters for specific properties) so that if there is a requirement to create a JDBC connection to the DB, they can do so. 

When [ConnectionFactory discovery](https://github.com/r2dbc/r2dbc-spi/issues/10) is merged maybe we could use that. 

- Code reuse. Current implementation of `R2dbcDatabaseContainer` is mostly copied from `JdbcDatabaseContainer`, and all concrete implementation would contain lots of duplicate code as well. To solve this we could either:
  - create another abstraction layer between `R2dbcDatabaseContainer` and `GenericDatabaseContainer`
  - pull as much code up into `GenericDatabaseContainer`  as possible add R2dbc compatibility at that level (seems unlikely)
  - maybe create some helper classes for the common code

Looking forward to your opinions, I'm happy to be devastated in a review

### CODE REVIEW

**Comment by mp911de** [comment]
Awesome news. You might want to wait until the configuration/discoverability PR (https://github.com/r2dbc/r2dbc-spi/pull/30) is merged. This will be an improvement to generic client bootstrap.

**Comment by Humblehound** [comment]
Thanks :) 

So that would totally be possible - I thought initially that it wouldn't work for the containers to use both JDBC and R2DBC at the same time and we would have to make the existing methods for getting `DataSource` throw and exception and immediately left the idea... but this is actually much simpler and should work with minimal impact on the existing design. The containers would use the existing infrastructure for setup/teardown etc and then just expose an R2DBC instance. Let me try that approach :)

File: settings.gradle [hunk 1]
```diff
@@ -35,4 +35,4 @@ project(':testcontainers').projectDir = "$rootDir/core" as File
 file('modules').eachDir { dir ->
     include dir.name
     project(":${dir.name}").projectDir = dir
-}
\ No newline at end of file
+}
```
[the remark this answers is not in the document]
Comment (Humblehound) [reply settings.gradle]: Is it feasible to rewrite history in PR's? I could just force push a cleaned version to make this thread cleaner, but I have no idea how this affects discussions like this.

[the remark this answers is not in the document]
Comment (bsideup) [reply settings.gradle]: don't worry, we will squash & merge it later, you can push as many commits as you want :)

**Comment by Humblehound** [comment]
This proved to be much simpler than I expected - I've reverted previous commits and just exposed a `ConnectionFactory` from R2DBC. I can see only two issues left:

- I've needed to add the milestone repository to each module for the project to compile
- The naming for all DB containers and packages includes `jdbc` which is slightly 

If you are ok with both I'd add missing documentation

**Comment by mp911de** [comment]
Connection Factory Discovery has now landed in R2DBC. This way, using code is not required to touch driver classes but rather can declare options that are used to determine an appropriate driver:

```java
ConnectionFactoryOptions options = builder()
    .option(DRIVER, "mssql") // alternative: postgresql
    .option(HOST, "…")
    .option(PORT, …)
    .option(USER, "…")
    .option(PASSWORD, "…")
    .option(DATABASE, "…")
    .build();

ConnectionFactory connectionFactory = ConnectionFactories.get(options);

Mono<Connection> connectionMono = factory.create();
```

**Comment by Humblehound** [comment]
While this syntax is easier to use for the end user, here I think that using specific implementations would be better. The ConnectionFactory needs to be returned for each driver:

```
            @Override
	    public ConnectionFactory getR2dbcConnectionFactory() {
	        return new PostgresqlConnectionFactory(PostgresqlConnectionConfiguration.builder()
	            .host(getContainerIpAddress())
	            .database(databaseName)
	            .password(password)
	            .username(username)
	            .port(getFirstMappedPort())
	            .build());
	    }
```

If were to use `ConnectionFactoryOptions` we would be getting rid of type safety.

**Comment by nebhale** [comment]
@Humblehound I’m not sure you lose any type safety with `ConnectionFactoryOptions`.  The option values are all strongly typed (`String`, `Integer`, etc.) and in your example, you’re returning the interface type `ConnectionFactory` rather than `PostgresqlConnectionFactory`.

**Comment by Humblehound** [comment]
@nebhale Indeed, but here I just don't see a benefit in passing a string containing a driver name to the `ConnectionFactoryOptions` so that it can determine what driver to use, when we know exactly what driver we need to use in each case. This would've been very useful if we needed to determine the driver in the generic container, but because we need to include a dependency on a specific driver for each supported DB module, we might as wall explicitly specify which driver to use. It just looks to me as a less complex solution as a whole. Unless I'm missing something here, like I mentioned I'm not very familiar with R2dbc source code.

**Comment by Humblehound** [comment]
But of course, it's your project so I'm happy to refactor to `ConnectionFactoryOptions`

**Comment by nebhale** [comment]
No, no I didn't want to be strongly one way or the other 😄  I just wanted to clarify the lack of type safety.  What I think @mp911de was getting at is that using `ConnectionFactories` would allow you to write a single implementation for all of the configuration (shared constants for `host`, `port`, `user`, `password`, and `database`) and then the _only_ set `driver` on a per-type basis.  If that level of sharing isn't desired then the driver-specific types are perfectly adequate.

**Comment by bsideup** [comment]
@Humblehound @nebhale I think the best end goal we could achieve is supporting R2DBC URI (like we do with JDBC) and implementing a fake, proxying R2DBC driver inside Testcontainers.

I don't think it makes sense to add R2DBC connection factory getter to the containers (because it is trivial to do on the user's side).

Also, R2DBC is not available in Maven Central yet and we can't depend on it directly, only with Gradle's `compileOnly` linking.

**Comment by nebhale** [comment]
> Also, R2DBC is not available in Maven Central yet and we can't depend on it directly, only with Gradle's compileOnly linking.

Already discussing internally.

**Comment by nebhale** [comment]
@kiview We'll be [adding URI-based configuration](https://github.com/r2dbc/r2dbc-spi/issues/37) (in a more structured way than JDBC) shortly.  To get to parity I figure you'd have a `getR2dbcUri()` method and the ability to turn an R2DBC URI transparently into a `ConnectionFactory` [in the same way that `DataSource`s are created](https://testcontainers.gitbook.io/workshop/step-4-your-first-testcontainers-integration).  There appears to be exposure of the JDBC Driver instance (`getJdbcDriverInstance()`) but I'm hoping that R2DBC is constructed in a way that doesn't make that required.

**Comment by Humblehound** [comment]
Perhaps I can work on the URI-based configuration in the meantime, unless you already started doing that. I assume that this PR will not be accepted at least until r2dbc is available from official repositories?

**Comment by nebhale** [comment]
The URI-based approach is specifically out of the current milestone (M7) which we're targeting for some time before the end of January. So while I encourage you to contribute to R2DBC, we're not going to merge any contribution before we start work on M8 simply to ensure we don't get scope creep.  Hope this doesn't turn you off too much.

**Comment by bsideup** [comment]
@Humblehound if we do it in a module (R2DBC url support), with `compileOnly` dependency on r2dbc - then it is okay to accept it before the release of R2DBC.

**Comment by Humblehound** [comment]
@bsideup Ok, that's good to know, I'll see if I can come up with a way to externalise it.
@nebhale Not an issue, I'm looking forward to working towards something more long-term. I'll follow up on this within the mailing group or on the r2dbc GitHub page.

**Comment by Humblehound** [comment]
I've refactored the code into a single module with support for postgres and mssql.

File: modules/r2dbc/src/main/java/org/testcontainers/r2dbc/R2dbcMSSQLServerContainer.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package org.testcontainers.r2dbc;
+
+import io.r2dbc.mssql.MssqlConnectionConfiguration;
+import io.r2dbc.mssql.MssqlConnectionFactory;
+import io.r2dbc.spi.ConnectionFactory;
+import org.testcontainers.containers.MSSQLServerContainer;
+
+/**
+ * @author humblehound
```
Comment (kiview) [inline modules/r2dbc/src/main/java/org/testcontainers/r2dbc/R2dbcMSSQLServerContainer.java]: Although we don't have a clear policy for `@author` tags in the project, I personally prefer to omit them in regards to shared code ownership.

File: modules/r2dbc/src/test/java/SimpleR2dbcMSSQLServerTest.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.testcontainers.r2dbc.R2dbcMSSQLServerContainer;
+
+import static org.rnorth.visibleassertions.VisibleAssertions.assertEquals;
+
+/**
+ * @author humblehound
+ */
+public class SimpleR2dbcMSSQLServerTest {
```
[the remark this answers is not in the document]
Comment (kiview) [reply modules/r2dbc/src/test/java/SimpleR2dbcMSSQLServerTest.java]: I agree, the tests should go into the corresponding package.

File: modules/r2dbc/src/test/java/SimpleR2dbcPostgreSQLTest.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import io.r2dbc.client.R2dbc;
+import org.junit.Test;
+import org.testcontainers.r2dbc.R2dbcPostgresContainer;
+
+import static org.rnorth.visibleassertions.VisibleAssertions.assertEquals;
+
+/**
+ * @author humblehound
+ */
+public class SimpleR2dbcPostgreSQLTest {
```
[the remark this answers is not in the document]
Comment (kiview) [reply modules/r2dbc/src/test/java/SimpleR2dbcPostgreSQLTest.java]: Same as with the other test.

File: modules/r2dbc/src/main/java/org/testcontainers/r2dbc/R2dbcMSSQLServerContainer.java (hunk 2 above)
Comment (Humblehound) [reply modules/r2dbc/src/main/java/org/testcontainers/r2dbc/R2dbcMSSQLServerContainer.java]: I don't feel strong about them either, just followed what I already found in the repo. Will remove

File: modules/r2dbc/src/test/java/SimpleR2dbcMSSQLServerTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Humblehound) [reply modules/r2dbc/src/test/java/SimpleR2dbcMSSQLServerTest.java]: missed that one ;)

File: modules/r2dbc-test/src/test/java/org/testcontainers/r2dbc/SimpleR2dbcMSSQLServerTest.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+public class SimpleR2dbcMSSQLServerTest {
+
+    @Rule
+    public R2dbcMSSQLServerContainer mssqlServer = new R2dbcMSSQLServerContainer();
+
+    @Test
+    public void testR2dbcConnectionSuccessful() {
+        R2dbc r2dbc = new R2dbc(mssqlServer.getR2dbcConnectionFactory());
+        Object resultSetInt =  r2dbc.inTransaction(handle -> handle.createQuery("SELECT 1").mapRow(row -> row.get(""))).blockFirst();
```
Comment (bsideup) [inline modules/r2dbc-test/src/test/java/org/testcontainers/r2dbc/SimpleR2dbcMSSQLServerTest.java]:
```suggestion
        Object resultSetInt = r2dbc.inTransaction(handle -> handle.createQuery("SELECT 1").mapRow(row -> row.get(""))).blockFirst();
```

File: modules/r2dbc-test/src/test/java/org/testcontainers/r2dbc/SimpleR2dbcPostgreSQLTest.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import org.junit.Test;
+import org.testcontainers.r2dbc.R2dbcPostgresContainer;
+
+import static org.rnorth.visibleassertions.VisibleAssertions.assertEquals;
+
+public class SimpleR2dbcPostgreSQLTest {
+
+    @Test
+    public void testR2dbcConnectionSuccessful() {
+        try (R2dbcPostgresContainer postgres = (R2dbcPostgresContainer) new R2dbcPostgresContainer()
```
Comment (bsideup) [inline modules/r2dbc-test/src/test/java/org/testcontainers/r2dbc/SimpleR2dbcPostgreSQLTest.java]:
```suggestion
        try (R2dbcPostgresContainer postgres = new R2dbcPostgresContainer()
```

File: modules/r2dbc/src/main/java/org/testcontainers/r2dbc/R2dbcMSSQLServerContainer.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package org.testcontainers.r2dbc;
+
+import io.r2dbc.mssql.MssqlConnectionConfiguration;
+import io.r2dbc.mssql.MssqlConnectionFactory;
+import io.r2dbc.spi.ConnectionFactory;
+import org.testcontainers.containers.MSSQLServerContainer;
+
+public class R2dbcMSSQLServerContainer extends MSSQLServerContainer implements R2dbcSupport {
```
Comment (bsideup) [inline modules/r2dbc/src/main/java/org/testcontainers/r2dbc/R2dbcMSSQLServerContainer.java]:
```suggestion
public class R2dbcMSSQLServerContainer extends MSSQLServerContainer<R2dbcMSSQLServerContainer> implements R2dbcSupport {
```

File: modules/r2dbc/src/main/java/org/testcontainers/r2dbc/R2dbcPostgresContainer.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package org.testcontainers.r2dbc;
+
+
+import io.r2dbc.postgresql.PostgresqlConnectionConfiguration;
+import io.r2dbc.postgresql.PostgresqlConnectionFactory;
+import io.r2dbc.spi.ConnectionFactory;
+import org.testcontainers.containers.PostgreSQLContainer;
+
+public class R2dbcPostgresContainer extends PostgreSQLContainer implements R2dbcSupport {
```
Comment (bsideup) [inline modules/r2dbc/src/main/java/org/testcontainers/r2dbc/R2dbcPostgresContainer.java]:
```suggestion
public class R2dbcPostgresContainer extends PostgreSQLContainer<R2dbcPostgresContainer> implements R2dbcSupport {
```

File: modules/r2dbc-test/src/test/java/org/testcontainers/r2dbc/SimpleR2dbcMSSQLServerTest.java (hunk 5 above)
Comment (Humblehound) [reply modules/r2dbc-test/src/test/java/org/testcontainers/r2dbc/SimpleR2dbcMSSQLServerTest.java]: done

**Comment by Humblehound** [comment]
I believe I've resolved all the conflicts and done all requested changes. I've also added a sample doc section. Anything else left to improve?

**Comment by Humblehound** [comment]
@bsideup @kiview I'd appreciate some feedback / action, so that this PR doesn't hang here forever. Thanks alot

**Comment by bsideup** [comment]
FYI https://github.com/r2dbc/r2dbc-spi/issues/37 was merged and now we can finally continue with the R2DBC support 🎉

**Comment by mp911de** [comment]
R2DBC 0.8.0.M8 is released and so URL formats are stable.

R2DBC Javadoc: https://r2dbc.io/spec/0.8.0.M8/api/

**Comment by rnorth** [comment]
@mp911de, @kiview and I met and discussed a couple of weeks ago. Sorry for the late writeup, but what we discussed was a plan to:

* With the new milestone release of R2DBC, introduce a `String getR2dbcUrl()` method for the _existing_ database container classes. As a simple string getter, we'd have no binary dependency on R2DBC, so this would be straightforward to add to our existing modules. (@bsideup I know you have some thoughts on this - let's discuss)

* Later, we'll do a separate piece of work to implement a proxying driver implementation, similar to our `ContainerDatabaseDriver` class and conforming to a R2DBC SPI. 

The latter would have a binary dependency on R2DBC and thus belongs in a new module.

As such, I think this particular PR can be trimmed a lot or replaced, unfortunately @Humblehound. Thank you for sparking the conversations, and sorry it's taken a long time so far.

**Comment by Humblehound** [comment]
No problem, I'm glad to have the problem resolved one way or another.

**Comment by daggerok** [comment]
hello, so what about testing r2dbc spring-boot apps with postgres in test-container?

**Pull request closed without merging**